h5py/tests/test_dataset.py

```diff
@@ -165,6 +166,63 @@ def test_empty_create_via_Empty_class(self):
         self.f.create_dataset('foo', data=h5py.Empty(dtype='f'))
         self.assertTrue(is_empty_dataspace(self.f['foo'].id))
 
+    def test_create_incompatible_data(self):
+        data = self.f.create_dataset('dset', (3,))
+        with self.assertRaises(ValueError):
+            self.f.create_dataset('bar', shape=4, data=data)
+
+
+class TestReadDirectly(BaseDataset):
+
+    """
+        Feature: Read data directly from Dataset into a Numpy array
+    """
+
+    def test_read_direct(self):
+        dset = self.f.create_dataset("dset", (100,), dtype='int64')
+        source_dset = self.f.create_dataset("source_dset", (100,), dtype='int64')
+        dest_dset = self.f.create_dataset("dest_dset", (100,), dtype='int64')
+        empty_dset = self.f.create_dataset("edset", dtype='int64')
+
+        source_sel = sel.select((100,), ..., source_dset)
+        dest_sel = sel.select((100,), ..., dest_dset)
+        arr = np.zeros((100,), dtype='int32')
+
+        if arr.flags.c_contiguous:
+            # read empty dataset
+            with self.assertRaises(TypeError):
+                empty_dset.read_direct(arr, np.s_[0:10], np.s_[50:60])
+
+            dset.read_direct(arr, np.s_[0:10], np.s_[50:60])
+            self.assertEqual(dset.shape, (100,))
+
+            dset.read_direct(arr, source_sel, dest_sel)
+            self.assertEqual(dset.shape, (100,))
+
+class TestWriteDirectly(BaseDataset):
+
+    """
+        Feature: Write Numpy array directly into Dataset
+    """
+
+    def test_write_direct(self):
+        dset = self.f.create_dataset('dset', (100,), dtype='int32')
+        empty_dset = self.f.create_dataset("edset", dtype='int64')
+
+        arr = np.ones((100,), dtype='int64')
+
+        if arr.flags.c_contiguous:
+            # write into empty dataset
+            # FIXME: write data into empty datase should be allowed
+            with self.assertRaises(TypeError):
+                empty_dset.write_direct(arr, np.s_[0:10], np.s_[50:60])
+
+            dset.write_direct(arr, np.s_[0:10], np.s_[50:60])
+            self.assertEqual(dset.shape, (100,))
+
+            dset.write_direct(arr)
+            self.assertEqual(dset.shape, (100,))
```

Comment:
These don't really make sense here either. `write_direct` doesn't affect the shape - only explicitly calling `dset.resize()` does that.

We could check for 1s in the relevant bit of the dataset after writing to it.